docs/proposals/004-initialization.md

```diff
@@ -0,0 +1,318 @@
+type is default-initializable. If the member type is not default-initializable and a default value isn't provided on the member, then such the constructor
+synthesis will fail and the constructor will not be added to the type. Failure to synthesis a constructor is not an error, and an error will appear
+if the user is trying to initialize a value of the type in question assuming such a constructor exist.
+
+Note that if every member of a struct contains a default expression, the synthesized `__init` method can be called with 0 arguments, however, this will not cause a variable declaration to be implicitly initialized. Implicit initialization is a backward compatibility feature that only work for user-defined `__init()` methods.
+
+### Single argument constructor call
+
+Call to a constructor with a single argument is always treated as a syntactic sugar of type cast:
+```csharp
+int x = int(1.0f); // is treated as (int) 1.0f;
+MyType y = MyType(arg); // is treated as (MyType)arg;
+MyType x = MyType(y); // equivalent to `x = y`.
+```
+
+The compiler will attempt to resolve all type casts using type coercion rules, if that failed, will fall back to resolve it as a constructor call.
+
+### Initialization List
+
+Slang allows initialization of a variable by assigning it with an initialization list. 
+Generally, Slang will always try to resolve initialization list coercion as if it is an explicit constructor invocation.
+For example, given:
+```csharp
+S obj = {1,2};
+```
+Slang will try to convert the code into:
+```csharp
+S obj = S(1,2);
+```
+
+Following the same logic, an empty initializer list will translate into a default-initialization:
+```csharp
+S obj = {};
+// equivalent to:
+S obj = S();
+```
+
+Note that initializer list of a single argument still translates directly into a constructor call and not a type cast. For example:
+```csharp
+void test()
+{
+  MyType t = {1};
+  // translates to:
+  // MyType t = MyType.__init(1);
+  // which is not
+  // MyType t = MyType(t)
+  // or
+  // MyType t = (MyType)t;
+}
+```
+
+If the above code passes type check, then it will be used as the way to initialize `obj`.
+
+If the above code does not pass type check, Slang continues to check if `S` meets the standard of a "legacy C-style struct` type.
+A type is a "legacy C-Style struct" if all of the following conditions are met:
+- It is a user-defined struct type or a basic scalar, vector or matrix type, e.g. `int`, `float4x4`.
+- It does not contain any explicit constructors defined by the user.
+- All its members have higher or equal visibility than the type.
+- All its members are legacy C-Style structs or arrays of legacy C-style structs.
+In such case, we perform a legacy "read data" style consumption of the initializer list, so that the following behavior is valid:
+
+```csharp
+struct Inner { int x; int y; };
+struct Outer { Inner i; Inner j; }
+
+Outer o = {1, 2, 3}; // Initializes `o` into `{ Inner{1,2}, Inner{3,0} }`.
```

Comment:
Is this already supported now?

Reply:
Yes, this is already implemented in the legacy logic.